File: Utilities/Scripts/Google/SearchConsole/searchConsole.py
from googleapiclient.discovery import build
from google.oauth2.credentials import Credentials
import sys
import json
from datetime import datetime, timedelta

def get_google_search_console_metrics(access_token, site_url, start_date, end_date):
    creds = Credentials(token=access_token)

    service = build('searchconsole', 'v1', credentials=creds)

    response = service.searchanalytics().query(
        siteUrl=site_url,
        body={
            'startDate': start_date,
            'endDate': end_date,
            'rowLimit': 1
        }
    ).execute()

    row = response.get('rows', [{}])[0]
    metrics = {
        "total_clicks": row.get('clicks', 0),
        "total_impressions": row.get('impressions', 0),
        "average_ctr": row.get('ctr', 0),
        "average_position": row.get('position', 0)
    }

    return metrics

def calculate_change(previous, current):
    if previous == 0:
        return 0 if current == 0 else 0
    return ((current - previous) / previous) * 100

if __name__ == "__main__":
    access_token = sys.argv[1]
    site_url = sys.argv[2]
    start_date = sys.argv[3]
    end_date = sys.argv[4]

    start_date_obj = datetime.strptime(start_date, '%Y-%m-%d')
    end_date_obj = datetime.strptime(end_date, '%Y-%m-%d')

    delta = end_date_obj - start_date_obj
    previous_start_date = (start_date_obj - delta).strftime('%Y-%m-%d')
    previous_end_date = (end_date_obj - delta).strftime('%Y-%m-%d')

    current_metrics = get_google_search_console_metrics(access_token, site_url, start_date, end_date)
    previous_metrics = get_google_search_console_metrics(access_token, site_url, previous_start_date, previous_end_date)

    clicks_change = calculate_change(previous_metrics["total_clicks"], current_metrics["total_clicks"])
    impressions_change = calculate_change(previous_metrics["total_impressions"], current_metrics["total_impressions"])
    ctr_change = calculate_change(previous_metrics["average_ctr"], current_metrics["average_ctr"])
    position_change = calculate_change(previous_metrics["average_position"], current_metrics["average_position"])

    result = {
        "total_clicks": current_metrics["total_clicks"],
        "total_impressions": current_metrics["total_impressions"],
        "average_ctr": current_metrics["average_ctr"],
        "average_position": current_metrics["average_position"],
        "clicks_change": clicks_change,
        "impressions_change": impressions_change,
        "ctr_change": ctr_change,
        "position_change": position_change
    }

    print(json.dumps(result, indent=2, ensure_ascii=False))

File: Utilities/Utilities/GoogleData/SearchConsole/SearchConsole.cs
using Google.Apis.Auth.OAuth2;
using Google.Apis.SearchConsole.v1.Data;
using Google.Apis.SearchConsole.v1;
using Google.Apis.Services;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Utilities.Utilities.GoogleData.SearchConsole
{
    public class SearchConsoleMetrics
    {
        public int TotalClicks { get; set; }
        public int TotalImpressions { get; set; }
        public double AverageCtr { get; set; }
        public double AveragePosition { get; set; }
        public double ClicksChange { get; set; }
        public double ImpressionsChange { get; set; }
        public double CtrChange { get; set; }
        public double PositionChange { get; set; }
    }

    public class SearchConsole
    {
        public SearchConsoleService CreateSearchConsoleService(string accessToken)
        {
            var credential = GoogleCredential.FromAccessToken(accessToken);
            var service = new SearchConsoleService(new BaseClientService.Initializer()
            {
                HttpClientInitializer = credential,
                ApplicationName = "Google Search Console API",
            });
            return service;
        }

        public Dictionary<string, object> GetGoogleSearchConsoleMetrics(SearchConsoleService service, string siteUrl, string startDate, string endDate)
        {
            var request = service.Searchanalytics.Query(new SearchAnalyticsQueryRequest
            {
                StartDate = startDate,
                EndDate = endDate,
                RowLimit = 1
            }, siteUrl);

            var response = request.Execute();

			var row = response.Rows?.FirstOrDefault();

			if (row == null)
			{
				return new Dictionary<string, object>
		        {
			        { "total_clicks", 0 },
			        { "total_impressions", 0 },
			        { "average_ctr", 0 },
			        { "average_position", 0 }
		        };
			}

			var metrics = new Dictionary<string, object>
	        {
		        { "total_clicks", row.Clicks ?? 0 },
		        { "total_impressions", row.Impressions ?? 0 },
		        { "average_ctr", row.Ctr ?? 0 },
		        { "average_position", row.Position ?? 0 }
	        };

			return metrics;
        }

        public double CalculateChange(double previous, double current)
        {
            if (previous == 0)
            {
                return current == 0 ? 0 : 0;
            }
            return (current - previous) / previous * 100;
        }

        public SearchConsoleMetrics GetSearchConsoleMetrics(string accessToken, string siteUrl, string startDate, string endDate)
        {
            var service = CreateSearchConsoleService(accessToken);

            var currentMetrics = GetGoogleSearchConsoleMetrics(service, siteUrl, startDate, endDate);

            var startDateObj = DateTime.Parse(startDate);
            var endDateObj = DateTime.Parse(endDate);
            var delta = endDateObj - startDateObj;
            var previousStartDate = startDateObj.AddDays(-delta.Days).ToString("yyyy-MM-dd");
            var previousEndDate = endDateObj.AddDays(-delta.Days).ToString("yyyy-MM-dd");

            var previousMetrics = GetGoogleSearchConsoleMetrics(service, siteUrl, previousStartDate, previousEndDate);

            var clicksChange = CalculateChange(Convert.ToDouble(previousMetrics["total_clicks"]), Convert.ToDouble(currentMetrics["total_clicks"]));
            var impressionsChange = CalculateChange(Convert.ToDouble(previousMetrics["total_impressions"]), Convert.ToDouble(currentMetrics["total_impressions"]));
            var ctrChange = CalculateChange(Convert.ToDouble(previousMetrics["average_ctr"]), Convert.ToDouble(currentMetrics["average_ctr"]));
            var positionChange = CalculateChange(Convert.ToDouble(previousMetrics["average_position"]), Convert.ToDouble(currentMetrics["average_position"]));

            var result = new SearchConsoleMetrics
            {
                TotalClicks = Convert.ToInt32(currentMetrics["total_clicks"]),
                TotalImpressions = Convert.ToInt32(currentMetrics["total_impressions"]),
                AverageCtr = Convert.ToDouble(currentMetrics["average_ctr"]),
                AveragePosition = Convert.ToDouble(currentMetrics["average_position"]),
                ClicksChange = clicksChange,
                ImpressionsChange = impressionsChange,
                CtrChange = ctrChange,
                PositionChange = positionChange
            };

            return result;
        }
    }
}
